File: ios/DailyMemory/Data/Remote/ImageAnalysisService.swift
import Foundation
import UIKit

/// Service for analyzing images using Vision AI (OpenAI GPT-4o Vision)
/// Extracts objects, scene descriptions, text (OCR), and face counts
actor ImageAnalysisService {
    static let shared = ImageAnalysisService()

    private let apiClient = APIClient.shared

    private init() {}

    // MARK: - Analysis

    /// Analyze an image via backend API (falls back to simulation if offline)
    func analyzeImage(_ image: UIImage) async -> Result<ImageAnalysisResult, Error> {
        // Resize and encode image
        let maxDimension: CGFloat = 1024
        let resized = resizeImage(image, maxDimension: maxDimension)
        guard let imageData = resized.jpegData(compressionQuality: 0.7) else {
            return .success(simulateAnalysis(image))
        }
        let base64 = imageData.base64EncodedString()

        do {
            let response: ImageAnalysisResponse = try await apiClient.post("ai/analyze-image", body: ["imageBase64": base64])
            let result = ImageAnalysisResult(
                objects: response.objects,
                scene: response.scene,
                text: response.text,
                faces: response.faces,
                description: response.description,
                suggestedTags: response.suggestedTags
            )
            return .success(result)
        } catch {
            return .success(simulateAnalysis(image))
        }
    }

    private func resizeImage(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
        let size = image.size
        guard size.width > maxDimension || size.height > maxDimension else { return image }
        let ratio = min(maxDimension / size.width, maxDimension / size.height)
        let newSize = CGSize(width: size.width * ratio, height: size.height * ratio)
        UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
        image.draw(in: CGRect(origin: .zero, size: newSize))
        let resized = UIGraphicsGetImageFromCurrentImageContext() ?? image
        UIGraphicsEndImageContext()
        return resized
    }

    /// Analyze image from file URL
    func analyzeImage(from url: URL) async -> Result<ImageAnalysisResult, Error> {
        guard let image = UIImage(contentsOfFile: url.path) else {
            return .failure(ImageAnalysisError.cannotLoadImage)
        }
        return await analyzeImage(image)
    }

    /// Analyze image from Data
    func analyzeImage(from data: Data) async -> Result<ImageAnalysisResult, Error> {
        guard let image = UIImage(data: data) else {
            return .failure(ImageAnalysisError.cannotLoadImage)
        }
        return await analyzeImage(image)
    }

    // MARK: - Legacy (removed - now uses backend API)

    private func buildAnalysisPrompt() -> String {
        """
        Analyze this image and extract structured information.
        Return a JSON object with these fields:
        - objects: array of objects/items visible in the image (e.g., "person", "dog", "cake", "car")
        - scene: brief description of the scene/setting (e.g., "birthday party indoor", "beach sunset")
        - text: any visible text in the image (OCR), or null if none
        - faces: number of human faces visible (integer)
        - description: a brief one-sentence description of what's happening in the image
        - suggestedTags: array of relevant tags for categorizing this image (e.g., "family", "celebration", "travel")

        Respond ONLY with valid JSON, no additional text or markdown.
        """
    }

    // MARK: - Response Parsing

    private func parseAnalysisResponse(_ data: Data) throws -> ImageAnalysisResult {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ImageAnalysisError.invalidResponse
        }

        // Extract content from OpenAI response
        guard let choices = json["choices"] as? [[String: Any]],
              let firstChoice = choices.first,
              let message = firstChoice["message"] as? [String: Any],
              let content = message["content"] as? String else {
            throw ImageAnalysisError.invalidResponse
        }

        // Extract JSON from content
        guard let jsonStart = content.firstIndex(of: "{"),
              let jsonEnd = content.lastIndex(of: "}") else {
            throw ImageAnalysisError.invalidResponse
        }

        let jsonString = String(content[jsonStart...jsonEnd])
        guard let jsonData = jsonString.data(using: .utf8) else {
            throw ImageAnalysisError.invalidResponse
        }

        return try JSONDecoder().decode(ImageAnalysisResult.self, from: jsonData)
    }

    // MARK: - Simulation

    /// Simulate image analysis for testing/demo purposes
    private func simulateAnalysis(_ image: UIImage) -> ImageAnalysisResult {
        // Basic simulation based on image characteristics
        let width = image.size.width
        let height = image.size.height
        let aspectRatio = width / height

        var objects: [String] = []
        var scene = "general scene"
        var suggestedTags: [String] = []

        // Simulate based on aspect ratio
        if aspectRatio > 1.5 {
            // Wide/landscape image - likely outdoor or scenery
            scene = "outdoor landscape"
            objects = ["sky", "nature"]
            suggestedTags = ["outdoor", "scenery", "landscape"]
        } else if aspectRatio < 0.7 {
            // Tall/portrait image - likely portrait photo
            scene = "portrait photo"
            objects = ["person"]
            suggestedTags = ["portrait", "people"]
        } else {
            // Square-ish - could be anything
            scene = "casual photo"
            objects = ["person", "background"]
            suggestedTags = ["casual", "moment"]
        }

        return ImageAnalysisResult(
            objects: objects,
            scene: scene,
            text: nil,
            faces: 1,
            description: "A \(scene) captured in this photo",
            suggestedTags: suggestedTags
        )
    }

    // MARK: - Helpers

    private func resizeImageIfNeeded(_ image: UIImage, maxSize: CGFloat) -> UIImage {
        let width = image.size.width
        let height = image.size.height

        if width <= maxSize && height <= maxSize {
            return image
        }

        let ratio = min(maxSize / width, maxSize / height)
        let newSize = CGSize(width: width * ratio, height: height * ratio)

        let renderer = UIGraphicsImageRenderer(size: newSize)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: newSize))
        }
    }
}

// MARK: - Result Model

/// Result of image analysis
struct ImageAnalysisResult: Codable, Equatable {
    let objects: [String]
    let scene: String
    let text: String?
    let faces: Int
    let description: String
    let suggestedTags: [String]

    init(
        objects: [String] = [],
        scene: String = "",
        text: String? = nil,
        faces: Int = 0,
        description: String = "",
        suggestedTags: [String] = []
    ) {
        self.objects = objects
        self.scene = scene
        self.text = text
        self.faces = faces
        self.description = description
        self.suggestedTags = suggestedTags
    }
}

// MARK: - Errors

enum ImageAnalysisError: LocalizedError {
    case cannotLoadImage
    case cannotEncodeImage
    case invalidResponse
    case apiError(statusCode: Int)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .cannotLoadImage:
            return "Cannot load image"
        case .cannotEncodeImage:
            return "Cannot encode image for API"
        case .invalidResponse:
            return "Invalid response from Vision API"
        case .apiError(let statusCode):
            return "API error: HTTP \(statusCode)"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        }
    }
}

File: android/app/src/main/java/com/dailymemory/data/remote/ImageAnalysisService.kt
package com.dailymemory.data.remote

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.util.Base64
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Service for analyzing images using Vision AI (OpenAI GPT-4o Vision)
 * Extracts objects, scene descriptions, text (OCR), and face counts
 */
@Singleton
class ImageAnalysisService @Inject constructor(
    @ApplicationContext private val context: Context
) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(60, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .build()

    // API Configuration
    private var apiKey: String = ""

    /**
     * Configure the service with OpenAI API key
     */
    fun configure(apiKey: String) {
        this.apiKey = apiKey
    }

    /**
     * Analyze an image and extract information
     */
    suspend fun analyzeImage(bitmap: Bitmap): Result<ImageAnalysisResult> = withContext(Dispatchers.IO) {
        if (apiKey.isEmpty()) {
            // Return simulated result when no API key is configured
            return@withContext Result.success(simulateAnalysis(bitmap))
        }

        try {
            val result = analyzeWithOpenAI(bitmap)
            Result.success(result)
        } catch (e: Exception) {
            // Fallback to simulation on error
            Result.success(simulateAnalysis(bitmap))
        }
    }

    /**
     * Analyze image from URI
     */
    suspend fun analyzeImage(uri: Uri): Result<ImageAnalysisResult> = withContext(Dispatchers.IO) {
        try {
            val inputStream = context.contentResolver.openInputStream(uri)
                ?: return@withContext Result.failure(ImageAnalysisException("Cannot open image"))

            val bitmap = BitmapFactory.decodeStream(inputStream)
            inputStream.close()

            if (bitmap == null) {
                return@withContext Result.failure(ImageAnalysisException("Cannot decode image"))
            }

            analyzeImage(bitmap)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Analyze image from file path
     */
    suspend fun analyzeImage(file: File): Result<ImageAnalysisResult> = withContext(Dispatchers.IO) {
        try {
            val bitmap = BitmapFactory.decodeFile(file.absolutePath)
                ?: return@withContext Result.failure(ImageAnalysisException("Cannot decode image"))

            analyzeImage(bitmap)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    private suspend fun analyzeWithOpenAI(bitmap: Bitmap): ImageAnalysisResult {
        // Resize image if needed and convert to base64
        val resizedBitmap = resizeBitmapIfNeeded(bitmap, 1024)
        val base64Image = bitmapToBase64(resizedBitmap)

        if (resizedBitmap != bitmap) {
            resizedBitmap.recycle()
        }

        val requestBody = JSONObject().apply {
            put("model", "gpt-4o")
            put("messages", JSONArray().apply {
                put(JSONObject().apply {
                    put("role", "user")
                    put("content", JSONArray().apply {
                        put(JSONObject().apply {
                            put("type", "text")
                            put("text", buildAnalysisPrompt())
                        })
                        put(JSONObject().apply {
                            put("type", "image_url")
                            put("image_url", JSONObject().apply {
                                put("url", "data:image/jpeg;base64,$base64Image")
                                put("detail", "low")
                            })
                        })
                    })
                })
            })
            put("max_tokens", 500)
            put("temperature", 0.3)
        }.toString()

        val request = Request.Builder()
            .url("https://api.openai.com/v1/chat/completions")
            .addHeader("Content-Type", "application/json")
            .addHeader("Authorization", "Bearer $apiKey")
            .post(requestBody.toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: throw ImageAnalysisException("Empty response")

        if (!response.isSuccessful) {
            throw ImageAnalysisException("API error: HTTP ${response.code}")
        }

        return parseAnalysisResponse(responseBody)
    }

    private fun buildAnalysisPrompt(): String {
        return """
            Analyze this image and extract structured information.
            Return a JSON object with these fields:
            - objects: array of objects/items visible in the image (e.g., "person", "dog", "cake", "car")
            - scene: brief description of the scene/setting (e.g., "birthday party indoor", "beach sunset")
            - text: any visible text in the image (OCR), or null if none
            - faces: number of human faces visible (integer)
            - description: a brief one-sentence description of what's happening in the image
            - suggestedTags: array of relevant tags for categorizing this image (e.g., "family", "celebration", "travel")

            Respond ONLY with valid JSON, no additional text or markdown.
        """.trimIndent()
    }

    private fun parseAnalysisResponse(responseBody: String): ImageAnalysisResult {
        val json = JSONObject(responseBody)

        // Extract content from OpenAI response
        val choices = json.optJSONArray("choices")
            ?: throw ImageAnalysisException("Invalid response: no choices")
        val firstChoice = choices.optJSONObject(0)
            ?: throw ImageAnalysisException("Invalid response: no first choice")
        val message = firstChoice.optJSONObject("message")
            ?: throw ImageAnalysisException("Invalid response: no message")
        val content = message.optString("content")
            ?: throw ImageAnalysisException("Invalid response: no content")

        // Extract JSON from content
        val jsonStart = content.indexOf("{")
        val jsonEnd = content.lastIndexOf("}") + 1
        if (jsonStart < 0 || jsonEnd <= jsonStart) {
            throw ImageAnalysisException("Invalid response: no JSON found")
        }

        val jsonString = content.substring(jsonStart, jsonEnd)
        val resultJson = JSONObject(jsonString)

        // Parse objects
        val objects = mutableListOf<String>()
        resultJson.optJSONArray("objects")?.let { arr ->
            for (i in 0 until arr.length()) {
                objects.add(arr.getString(i))
            }
        }

        // Parse suggested tags
        val suggestedTags = mutableListOf<String>()
        resultJson.optJSONArray("suggestedTags")?.let { arr ->
            for (i in 0 until arr.length()) {
                suggestedTags.add(arr.getString(i))
            }
        }

        return ImageAnalysisResult(
            objects = objects,
            scene = resultJson.optString("scene", ""),
            text = resultJson.optString("text").takeIf { it.isNotEmpty() && it != "null" },
            faces = resultJson.optInt("faces", 0),
            description = resultJson.optString("description", ""),
            suggestedTags = suggestedTags
        )
    }

    /**
     * Simulate image analysis for testing/demo purposes
     */
    private fun simulateAnalysis(bitmap: Bitmap): ImageAnalysisResult {
        val width = bitmap.width
        val height = bitmap.height
        val aspectRatio = width.toFloat() / height.toFloat()

        var objects: List<String>
        var scene: String
        var suggestedTags: List<String>

        // Simulate based on aspect ratio
        when {
            aspectRatio > 1.5f -> {
                // Wide/landscape image - likely outdoor or scenery
                scene = "outdoor landscape"
                objects = listOf("sky", "nature")
                suggestedTags = listOf("outdoor", "scenery", "landscape")
            }
            aspectRatio < 0.7f -> {
                // Tall/portrait image - likely portrait photo
                scene = "portrait photo"
                objects = listOf("person")
                suggestedTags = listOf("portrait", "people")
            }
            else -> {
                // Square-ish - could be anything
                scene = "casual photo"
                objects = listOf("person", "background")
                suggestedTags = listOf("casual", "moment")
            }
        }

        return ImageAnalysisResult(
            objects = objects,
            scene = scene,
            text = null,
            faces = 1,
            description = "A $scene captured in this photo",
            suggestedTags = suggestedTags
        )
    }

    private fun resizeBitmapIfNeeded(bitmap: Bitmap, maxSize: Int): Bitmap {
        val width = bitmap.width
        val height = bitmap.height

        if (width <= maxSize && height <= maxSize) {
            return bitmap
        }

        val ratio = minOf(maxSize.toFloat() / width, maxSize.toFloat() / height)
        val newWidth = (width * ratio).toInt()
        val newHeight = (height * ratio).toInt()

        return Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true)
    }

    private fun bitmapToBase64(bitmap: Bitmap): String {
        val outputStream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 80, outputStream)
        val byteArray = outputStream.toByteArray()
        return Base64.encodeToString(byteArray, Base64.NO_WRAP)
    }
}

/**
 * Result of image analysis
 */
data class ImageAnalysisResult(
    val objects: List<String> = emptyList(),
    val scene: String = "",
    val text: String? = null,
    val faces: Int = 0,
    val description: String = "",
    val suggestedTags: List<String> = emptyList()
) {
    /**
     * Check if people were detected
     */
    val hasPeople: Boolean
        get() = faces > 0 || objects.any { it.lowercase().contains("person") }

    /**
     * Check if text was detected
     */
    val hasText: Boolean
        get() = !text.isNullOrEmpty()
}

/**
 * Exception for image analysis errors
 */
class ImageAnalysisException(message: String) : Exception(message)
